Source/Core/Core/HW/GCMemcard/GCMemcardUtils.cpp

```diff
@@ -0,0 +1,51 @@
+#include "Core/HW/GCMemcard/GCMemcardUtils.h"
+
+#include "Core/HW/GCMemcard/GCMemcard.h"
+
+namespace Memcard
+{
+bool HasSameIdentity(const DEntry& lhs, const DEntry& rhs)
+{
+  // The Gamecube BIOS identifies two files as being 'the same' (that is, disallows copying) when
+  // the full array of all of m_gamecode, m_makercode, and m_filename match between then.
+  // However, the m_filename part is typically a nullterminated string and should probably be
+  // treated as such, as some (most? all?) games identify their saves regardless of what bytes
+  // appear after the first null byte.
+
+  // That said, the BIOS behavior of doing a full compare of m_filename seems incorrect, as the
+  // following chain of actions fails with a rather vague 'The data may not have been copied.'
+  // error message:
+  // - Have two memory cards with one save file each.
+  // - The two save files should have identical gamecode and makercode, as well as an equivalent
+  //   filename up until and including the first null byte.
+  // - On Card A have all remaining bytes of the filename be null also.
+  // - On Card B have at least one of the remaining bytes be non-null.
+  // - Copy the file on Card B to Card A.
+  // The BIOS will abort halfway through the copy process and declare Card B as unusable until you
+  // eject and reinsert it, and leave a "Broken File000" file on Card A, though note that the file
+  // is not visible and will be cleaned up when reinserting the card while still within the BIOS.
+  // Also, either during or after the copy process, the bytes after the first null are 'fixed' to
+  // null, which is presumably why the copy process ends up failing as Card A would then have two
+  // identical files. For reference, the Wii System Menu behaves exactly the same.
+
+  // With all that in mind, I think we can safely assume that all bytes after the first null are to
+  // be ignored, and that m_filename is supposed to be a nullterminated string.
+
+  if (lhs.m_gamecode != rhs.m_gamecode)
+    return false;
+  if (lhs.m_makercode != rhs.m_makercode)
+    return false;
+
+  for (size_t i = 0; i < lhs.m_filename.size(); ++i)
```

Comment:
Should this perhaps be `i < std::min(lhs.m_filename.size(), rhs.m_filename.size())`? Or do we make sure that the `lhs` filename is shorter than (or of equal length to) the `rhs` filename?

Reply:
Scratch that, I just noticed this is an array and not an actual string.